File: 12-week-9--searching-and-sorting-continued/01-day-3-to-4--tree-traversal-in-order/javascript/tree_traversal_inorder.js
class Node {
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}

function treeTraversalInorder(root) {
  // type your code here
}

if (require.main === module) {
  // add your own tests in here
  let root = new Node(2, new Node(-10), new Node(20));
  console.log("Expecting: [-10, 2, 20]");
  console.log(treeTraversalInorder(root));

  console.log("");

  root = new Node(10, new Node(0, null, new Node(5)), new Node(20, null, new Node(30)));
  console.log("Expecting: [0, 5, 10, 20, 30] ");
  console.log(treeTraversalInorder(root));
}

module.exports = {
  Node,
  treeTraversalInorder
};

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 12-week-9--searching-and-sorting-continued/01-day-3-to-4--tree-traversal-in-order/ruby/tree_traversal_inorder.rb
class Node
  attr_accessor :value, :left, :right

  def initialize(value, left = nil, right = nil)
    @value = value
    @left = left
    @right = right
  end
end

def tree_traversal_inorder(root)
  # type your code in here
end

if __FILE__ == $PROGRAM_NAME
  root = Node.new(2, Node.new(-10), Node.new(20))
  puts "Expecting: [-10, 2, 20]"
  print tree_traversal_inorder(root)
  puts

  puts

  root = Node.new(10, Node.new(0, nil, Node.new(5)), Node.new(20, nil, Node.new(30)))
  puts "Expecting: [0, 5, 10, 20, 30] "
  print tree_traversal_inorder(root)
  puts

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
